electrumsv/network.py

```diff
@@ -1516,77 +607,102 @@ def update_api_servers_for_wallet(self,
 
     def add_wallet(self, wallet: "Wallet") -> None:
         """ This wallet has been loaded and is now using this network. """
+        self._wallets.add(wallet)
         self._register_api_servers_for_wallet(wallet)
-        app_state.async_.spawn(self._wallet_jobs.put, ('add', wallet))
 
     def remove_wallet(self, wallet: "Wallet") -> List[NetworkServerState]:
         """ This wallet has been unloaded and is no longer using this network. """
+        self._wallets.remove(wallet)
         updated_states = self._unregister_all_api_servers_for_wallet(wallet)
-        app_state.async_.spawn(self._wallet_jobs.put, ('remove', wallet))
         return updated_states
 
     def chain(self) -> Optional[Chain]:
-        main_session = self.main_session()
-        if main_session:
-            return main_session.chain
         return cast(Headers, app_state.headers).longest_chain()
 
-    def get_local_height(self) -> int:
-        chain = self.chain()
-        # This can be called from network_dialog.py when there is no chain
-        return cast(int, chain.height) if chain else 0
+    async def force_connect_header(self, raw_header: bytes, height: int) -> None:
+        try:
+            assert app_state.headers is not None
+            app_state.headers.connect(raw_header)
+        except MissingHeader:
+            await self._fetch_missing_headers_at_heights([height])
+
+    async def _fetch_missing_headers_at_heights(self, heights: Iterable[int]) -> Dict[int, Header]:
```

Comment:
Delete this function and never make anything like it. There are no gaps in the headers, and this starts creating gaps and this means it creates the problem it solves. We should never fetch arbitrary headers.